webapi/src/main/scala/org/knora/webapi/slice/admin/repo/service/KnoraUserGroupRepoLive.scala

```diff
@@ -17,48 +18,72 @@ import zio.ZIO
 import org.knora.webapi.slice.admin.domain.model.GroupIri
+import org.knora.webapi.slice.admin.domain.model.GroupStatus
 import org.knora.webapi.slice.admin.domain.model.KnoraProject.ProjectIri
-import org.knora.webapi.slice.admin.domain.model.KnoraUserGroup
+import org.knora.webapi.slice.admin.domain.model.KnoraUserGroup.Conversions.*
+import org.knora.webapi.slice.admin.domain.model._
 import org.knora.webapi.slice.admin.domain.service.KnoraUserGroupRepo
+import org.knora.webapi.slice.admin.repo.rdf.RdfConversions.projectIriConverter
 import org.knora.webapi.slice.admin.repo.rdf.Vocabulary
 import org.knora.webapi.slice.common.repo.rdf.RdfResource
 import org.knora.webapi.store.triplestore.api.TriplestoreService
 import org.knora.webapi.store.triplestore.api.TriplestoreService.Queries.Construct
+import org.knora.webapi.store.triplestore.api.TriplestoreService.Queries.Update
 import org.knora.webapi.store.triplestore.errors.TriplestoreResponseException
+import org.knora.webapi.store.triplestore.errors.TriplestoreUnsupportedFeatureException
 
 final case class KnoraUserGroupRepoLive(triplestore: TriplestoreService) extends KnoraUserGroupRepo {
-
   override def findById(id: GroupIri): Task[Option[KnoraUserGroup]] = for {
-    model    <- triplestore.queryRdfModel(UserGroupQueries.findById(id))
-    resource <- model.getResource(id.value)
-    user     <- ZIO.foreach(resource)(toGroup)
-  } yield user
+    model     <- triplestore.queryRdfModel(KnoraUserGroupQueries.findById(id))
+    resource  <- model.getResource(id.value)
+    userGroup <- ZIO.foreach(resource)(toGroup)
+  } yield userGroup
 
   override def findAll(): Task[List[KnoraUserGroup]] = for {
-    model <- triplestore.queryRdfModel(UserGroupQueries.findAll)
+    model <- triplestore.queryRdfModel(KnoraUserGroupQueries.findAll)
     resources <-
-      model.getResourcesRdfType(OntologyConstants.KnoraAdmin.UserGroup).option.map(_.getOrElse(Iterator.empty))
+      model.getResourcesRdfType(KnoraAdmin.UserGroup).option.map(_.getOrElse(Iterator.empty))
     groups <- ZStream.fromIterator(resources).mapZIO(toGroup).runCollect
   } yield groups.toList
 
-  private def toGroup(resource: RdfResource): Task[KnoraUserGroup] = for {
-    id <-
-      resource.iri.flatMap(it => ZIO.fromEither(GroupIri.from(it.value))).mapError(TriplestoreResponseException.apply)
-    belongsToProject <- resource
-                          .getObjectIriOrFail(OntologyConstants.KnoraAdmin.BelongsToProject)
-                          .flatMap(it => ZIO.fromEither(ProjectIri.from(it.value)))
-                          .mapError(it => TriplestoreResponseException(it.toString))
-  } yield KnoraUserGroup(id, belongsToProject)
+  def save(userGroup: KnoraUserGroup): Task[KnoraUserGroup] =
+    for {
+      query <- findById(userGroup.id).map {
+                 case Some(_) => throw new TriplestoreUnsupportedFeatureException("Updating users is not supported.")
```

Comment:
**issue**: The `save` method suggests it could persisted a new or existing group but fails on update

I would suggest you implement this similar to the [save query in the user repo](https://github.com/dasch-swiss/dsp-api/blob/main/webapi/src/main/scala/org/knora/webapi/slice/admin/repo/service/KnoraUserRepoLive.scala#L180-L224). This query is able to create a new user or update an existing. There is also no need for `findById` pre-query as the save query does is in one go.

Reply:
👍

Reply:
Take a look, what do you think?

Reply:
Morning!